src/trace.rs

```diff
@@ -271,13 +282,32 @@ impl<S> BroadcastLayer<S> {
 
 impl<S> Layer<S> for BroadcastLayer<S>
 where
-    S: Subscriber + 'static,
+    S: Subscriber + 'static + for<'lookup> LookupSpan<'lookup>,
 {
-    fn on_event(&self, event: &Event<'_>, _ctx: Context<'_, S>) {
-        // Try buffering the event, and if we're not buffering anymore, try to send it along via the
-        // trace sender if it's been established.
-        if !try_buffer_event(event) {
-            try_broadcast_event(event);
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        // Pass along span fields if available
+        if let Some(s) = ctx.event_span(event) {
+            if let Some(fields) = s.extensions().get::<LogEvent>() {
+                // Try buffering the event, and if we're not buffering anymore, try to send it along via the
+                // trace sender if it's been established.
+                if !try_buffer_event(event, Some(fields)) {
+                    try_broadcast_event(event, Some(fields));
+                }
+            }
+        } else if !try_buffer_event(event, None) {
+            try_broadcast_event(event, None);
         }
     }
+
+    fn on_new_span(
+        &self,
+        attrs: &tracing_core::span::Attributes<'_>,
+        id: &tracing_core::span::Id,
+        ctx: Context<'_, S>,
+    ) {
+        let span = ctx.span(id).expect("span must already exist!");
+        let mut fields = LogEvent::default();
+        attrs.values().record(&mut fields);
+        span.extensions_mut().insert(fields);
+    }
```

Comment:
What does an actual event _look like_ with this change?

From what I can see, I would almost expect to see the component-specific fields duplicated, existing both at the root of the log event and within the nested "metadata" structure?

Reply:
I guess my worry is that we're actually filtering out the fields on the top level object when sending through the `datadog_logs` sink, but they would be present in something like, say, the `console` sink or some other logging sink?

Additionally, I'm a little squeamish that we're converting the event once (`LogEvent::from`) and then visiting it _again_ and adding that as span data, and then finally extracting those fields and modifying the event to add them as metadata. It feels like we could do this in a cleaner, more efficient way by reintroducing `MakeLogEvent` but have it explicitly add our `component_*` fields under `metadata` as it's visiting, so that we only do a single pass over an event to build our desired `LogEvent`, without duplicating things like level, module path, etc.

Reply:
👍 You can see how the metrics exclude non-component labels here: https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/metrics/label_filter.rs

👍 to walking up the spans to find the labels.